### DISCUSSION TITLE
Missing onError handler with real Error objects

### DISCUSSION THREAD

**Discussion by tivac** [statelyai/xstate#1530] [Q&A]
So I found a discussion on this issue already, https://github.com/davidkpiano/xstate/issues/1187

But @davidkpiano's last comment doesn't make a ton of sense to me, and there's a really simple repro that is a convention we've been using for a while already.

https://codesandbox.io/s/missing-onerror-handler-but-its-really-there-wn6pc?file=/src/index.js

Can anybody more familiar with this code explain to me why this structure is inadvisable (aside from the obvious fact that it makes xstate barf an error...). I'm trying to escalate the original `new Error("NUH UH")` up to the parent machine so it can reported to the user.

```js
invoke: {
    id: "promise",
    src: () => Promise.reject(new Error("NUH UH")),

    onError: {
        actions: escalate("OH NO"),
    },
},
```

I've attempted to do some digging backwards into why this fails but not having much luck as of yet.

**Comment by davidkpiano** [comment]
The `onDone` and `onError` properties are for _transitions_, not for handlers. Using TypeScript would be helpful here, since it would immediately tell you the error.

Here is the fixed version: https://codesandbox.io/s/missing-onerror-handler-but-its-really-there-forked-ul2om?file=/src/index.js

```diff
const parent = Machine({
  initial: "a",

  states: {
    a: {
      invoke: {
        id: "child",
        src: child,
-       onError: (_, event) => { ... },
+       onError: {
          actions: (_, event) => {
            console.log("Received error:", event.data);
          }
        }
      }
    }
  }
});
```

**Reply by tivac** [reply]
Fair callout, that was my attempt at a simplified repro of something we're seeing in our codebase that *does* use `onError` correctly. Guess I'll keep chasing down a simplified repro 😞

**Reply by tivac** [reply]
Figured out the repro after a bunch more digging. It has to do with a sibling invoked machine that has `autoForward` set on it.

https://codesandbox.io/s/missing-onerror-handler-but-its-really-there-wn6pc

If you disable `autoForward` on line 71 in the sandbox will fix the issue, and is potentially something we can do in our actual production machine but will require some testing to be sure.

**Reply by Andarist** [reply]
So `autoForward` literally does what it says - it forwards all events to a particular machine. And `onError`/`onDone` are just sugars for transitions, as mentioned by @davidkpiano in the other comment. So if you receive an "error event" (it's still an event!) to a parent and you auto-forward that to another child then it's complaining about a missing transition defined for such an event.

**Reply by tivac** [reply]
Yeah, the confusing part is this message:

> Missing onError handler and/or unhandled exception/promise rejection for invocation 'c'.

Technically yes, `c` did fire off the error via `escalate()` but the missing `onError` handler was actually way over in the `b` machine. There's no way that I can see to know that unless you get lucky like I did and happen to comment out the  parallel invoked machine causing the issue.

**Reply by tivac** [reply]
I also don't actually know *how* you could handle the error from `c` in the parallel invoked machine `b`? Is the only option to disable autoForward?
